### PR TITLE
Feature/aww 1247 dex database facilities

### PR DESCRIPTION
**Pull request by nemezxxx**

### CODE REVIEW

File: apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public long transactionID;
+    @Schema(name="sender's offer ID", description="Identificator of exchange offer from the side of sender")
+    public long senderOfferID; 
+    @Schema(name="receiver's offer ID", description="Identificator of exchange offer from the side of receiver")
+    public long receiverOfferID;   
+    @Schema(name="Exchange type", description="buy/sell = 0/1")
+    public byte senderOfferType; 
+    @Schema(name="Type of currency", description="eth/pax/apl = 0/1/2")
+    public byte senderOfferCurrency; 
+    @Schema(name="Amount of currency for exchagne", description="parameter in gwei units")
```
Comment (AndrewBoyarsky) [inline apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java]: Typo `exchagne`

File: apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public long senderOfferAmount;
+    @Schema(name="Type of pair currency", description="eth/pax/apl = 0/1/2")
+    public byte pairCurrency;
+    @Schema(name="Exchange rate", description="parameter in gwei units")    
+    public BigDecimal pairRate;
```
Comment (AndrewBoyarsky) [inline apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java]: I think, that better to use long or BigInteger for rates

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+package com.apollocurrency.aplwallet.apl.exchange.model;
+
+import com.apollocurrency.aplwallet.api.dto.DexTradeInfoDto;
+import java.math.BigDecimal;
+
+/**
+ * Data class for interaction with trade table in the database
+ * @author Serhiy Lymar
+ */
+public class DexTradeEntry {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java]: Maybe better to use lombok?

File: apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java (hunk 2 above)
Comment (nemezxxx) [reply apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java]: BigInteger is not suitable cause it is not supposed to work with fractions. For the purposes of different RATE arithmetic, BigDecimal is what the doctor ordered. For ETH or BTC, by the way, BigDecimal is used. For both of them.

File: apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java (hunk 1 above)
Comment (nemezxxx) [reply apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java]: As far as I'm concerned, there's no any Unit tests inside the apl-api project. Or at least, in no any neighboring classes are any tests available. I wish I could find any template for that purpose.

Comment (nemezxxx) [reply apl-api/src/main/java/com/apollocurrency/aplwallet/api/dto/DexTradeInfoDto.java]: As far as I'm concerned, there's no any Unit tests inside the apl-api project. Or at least, in no any neighboring classes are any tests available. I wish I could find any template for that purpose.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java (hunk 3 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java]: I can't agree more. It's already done, thnx

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 4]
```diff
@@ -784,8 +784,15 @@ protected int update(int nextUpdate) {
             case 297:
                 apply("CREATE TABLE IF NOT EXISTS dex_contract (db_id IDENTITY NOT NULL, offer_id BIGINT NOT NULL, " +
                         "counter_offer_id BIGINT NOT NULL, secret_hash CHAR(64) NOT NULL, height INT NOT NULL, latest BOOLEAN NOT NULL DEFAULT TRUE)");
-            case 298:
-                return 298;
+            case 298:
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: That will be conflicting when stage is merged into master. Just FYI

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/DexController.java [hunk 5]
```diff
@@ -546,6 +624,60 @@ public Response dexEthInfo(@Context SecurityContext securityContext) throws NotF
...
+    @Operation(summary = "test flooding",
+            description = "initiates flooding table with the test data",
+            tags = {"debug"},
+            responses = {
+                    @ApiResponse(responseCode = "200", description = "Successful execution",
+                            content = @Content(mediaType = "application/json",
+                                    schema = @Schema(implementation = ResponseDone.class)))
+            }
+    ) 
+    public ResponseDone startTestFlood(  ) {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/DexController.java]: Maybe create ordinary api tests and service tests

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    private long receiverOfferID;
+    private byte senderOfferType; 
+    private byte senderOfferCurrency; 
+    private long senderOfferAmount;
+    private byte pairCurrency;
+    private BigDecimal pairRate;
+    private Integer finishTime;
+    private Integer height; 
+
+    public DexTradeInfoDto toDto(){
```
Comment (alzinchenko) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java]: I guess it's better to move the DTO conversion into separated service that implements the core/rest/converter/Converter interface.  The model class hasn't to contain any translating logic.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntryMin.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+@Getter @Setter
+public class DexTradeEntryMin {    
+     private BigDecimal hi;
+     private BigDecimal low;
+     private BigDecimal open;
+     private BigDecimal close;
+
+public DexTradeInfoMinDto toDto(){
```
Comment (alzinchenko) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntryMin.java]: .... the same DTO conversion

**Comment by nemezxxx** [comment]
> Unit tests required

done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 4 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: that's aparent. Anyway, while developing and merging sharding-related branch, that version should be updated. We will cross that bridge when it comes to it.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/DexController.java (hunk 5 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/rest/endpoint/DexController.java]: Relocated it from there to unit tests, thanks.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java (hunk 6 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntry.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntryMin.java (hunk 7 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexTradeEntryMin.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java [hunk 8]
```diff
@@ -87,6 +100,17 @@ public DexOffer getOfferByTransactionId(Long transactionId){
     public DexOffer getOfferById(Long id){
         return dexOfferDao.getById(id);
     }
+    
+    @Transactional
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java]: readonly=true

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java [hunk 9]
```diff
@@ -58,23 +60,34 @@
...
+
+    private DexTradeDao dexTradeDao;
+
     private DexOfferTransactionCreator dexOfferTransactionCreator;
     private EpochTime timeService;
 
-
     @Inject
     public DexService(EthereumWalletService ethereumWalletService, DexOfferDao dexOfferDao, DexOfferTable dexOfferTable, TransactionProcessorImpl transactionProcessor,
+// <<<<<<< HEAD
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java]: Clear merge info

**Pull request merged**